### PR TITLE
Various LIR improvements

### PR DESCRIPTION
**Pull request by andrewvc**
Many refactors
Rebase off newer master
Fixed an issue with dangling leaf nodes.
Improved Diff Reports
Various pieces of code cleanup.
More consistent ID generation.

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java [hunk 1]
```diff
@@ -20,6 +21,9 @@ public BooleanExpression getBooleanExpression() {
 
     public IfVertex(SourceMetadata meta, BooleanExpression booleanExpression) {
         super(meta);
+        this.id = "if-" +
+                (meta != null ? Integer.toString(meta.hashCode()) : UUID.randomUUID().toString())
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java]: This is intended to provide a consistent ID for if conditions. Maybe this isn't necessary, but it seems to be useful. Ideally we'd have a user-specified ID here.

File: logstash-core/src/main/java/org/logstash/config/ir/SourceMetadata.java [hunk 2]
```diff
@@ -40,6 +48,10 @@ public SourceMetadata() {
         this.sourceText = null;
     }
 
+    public int hashCode() {
+        return Objects.hash(this.sourceFile, this.sourceLine, this.sourceColumn, this.sourceText);
+    }
+
     public String toString() {
         return sourceFile + ":" + sourceLine + ":" + sourceColumn + ":```\n" + sourceText + "\n```";
     }
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/SourceMetadata.java]: We have decided to extract the `@JsonProperty` Does this PR need to be updated with that logic or we want to make it another PR for that.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package org.logstash.config.ir.graph;
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java]: Thanks for extracting this into his own class, I know this is mostly for debugging but can we have 2-3 unit test to make sure we don't regress it? I am pretty sure we improve this classe for other stuff.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    public DiffResult(Collection<Edge> removedEdges, Collection<Edge> addedEdges, List<Vertex> removedVertices, List<Vertex> addedVertices) {
+        this.removedEdges = removedEdges;
+        this.addedEdges = addedEdges;
+        this.removedVertices = removedVertices;
+        this.addedVertices = addedVertices;
+    }
+
+    public String summary() {
+        String template = "(-%d,+%d Edges | -%d,+%d Vertices)";
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java]: Maybe move this to a constant?

File: logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java (hunk 1 above)
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java]: I think this implementation is OK for now.

The only down side I currently see the implementation is tied to the ruby language implementation.
Maybe we should hide that generation logic behind a method that we can call on both BooleanExpression and IfVertex this will give us more future flexibility? WDYT?

File: logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java [hunk 5]
```diff
@@ -59,6 +63,16 @@ public boolean acceptsOutgoingEdge(Edge e) {
...
 
+    @Override
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public String typeString() {
+        return "conditional";
+    }
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java]: Since the possibility are limited should we use an Enum here?

File: logstash-core/src/test/java/org/logstash/config/ir/IRHelpers.java [hunk 6]
```diff
@@ -55,4 +71,18 @@ public static SourceMetadata testMetadata() {
...
+                                    iPlugin(FILTER, "grok"),
+                                    iPlugin(FILTER, "kv")).toGraph();
+        Graph outputSection = iIf(eGt(eEventValue("[baz]"), eValue(1000)),
+                                    iComposeParallel(
+                                            iPlugin(OUTPUT, "s3"),
+                                            iPlugin(OUTPUT, "elasticsearch")),
+                                    iPlugin(OUTPUT, "stdout")).toGraph();
+
+        return new Pipeline(inputSection, filterSection, outputSection);
+    }
```
Comment (ph) [inline logstash-core/src/test/java/org/logstash/config/ir/IRHelpers.java]: Where is this `samplePipeline` used? I don't see any reference to it in the current code?

File: logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java (hunk 1 above)
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java]: @ph how is it tied to the ruby lang implementation?

One downside here is that the hashCode isn't really unique enough. Java hash codes aren't intended to prevent collisions. I should probably change that.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java (hunk 4 above)
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java]: There's no need in Java due to the [String Pool](http://stackoverflow.com/questions/2486191/what-is-the-java-string-pool-and-how-is-s-different-from-new-strings)

Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/DiffResult.java]: TIL

File: logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java (hunk 1 above)
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/IfVertex.java]: > booleanExpression.toRubyString()

`toRubyString` feel specific to ruby ?:)

**Comment by andrewvc** [comment]
@ph the ID generation here I came up with is really crappy. I'm going to instead generate real SHA256 hashed ids consistently from subparts of the tree. That way you can compare any two subtrees / graphs etc. It's just the right way to do this that isn't all hacky. Will take a little while longer.

File: logstash-core/src/test/java/org/logstash/config/ir/IRHelpers.java (hunk 6 above)
Comment (andrewvc) [inline logstash-core/src/test/java/org/logstash/config/ir/IRHelpers.java]: It wasn't used before but is now in a new test!

**Comment by andrewvc** [comment]
@ph this is ready for review again.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java [hunk 7]
```diff
@@ -1,182 +1,294 @@
...
     }
 
+    private void calculateRanks() {
+        vertexRanks = BreadthFirst.breadthFirst(this.getRoots()).vertexDistances;
+    }
+
+    public Integer rank(Vertex vertex) {
+        Integer rank = vertexRanks.get(vertex);
+        // This should never happen
+        if (rank == null) throw new RuntimeException("Attempted to get rank from vertex where it is not yet calculated!" + this);
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java]: missing space after quote?

File: logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java [hunk 8]
```diff
@@ -1,182 +1,294 @@
...
+            String joinedErrorMessageGroups = errorMessageGroups.collect(Collectors.joining("\n---\n"));
 
-    // Walk the graph, visiting each edge on it with the provided Consumer<Edge>
-    public void walkEdges(Consumer<Edge> consumer) {
-        // Avoid stream interface to avoid concurrency issues if a new root is added
-        // since streams are lazy you can't mutate the graph while walking it with a
-        // stream. Using this iterative approach we can allow such operations.
-        for (Vertex root : this.getRoots()) {
-            walkEdges(consumer, root);
+            throw new InvalidIRException("Some nodes on the graph are fully redundant!\n" + joinedErrorMessageGroups);
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java]: Thanks for all the efforts in the validation!

File: logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java [hunk 9]
```diff
@@ -341,21 +418,16 @@ public boolean isEmpty() {
...
 
     public Stream<Edge> edges() {
         return this.edges.stream();
     }
+
+    @Override
+    public String hashSource() {
+        return this.vertices.stream().map(Vertex::hashSource).sorted().collect(Collectors.joining("\n"));
+    }
 }
```
Comment (ph) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java]: I like the proposed style `getRoots` vs `roots` for streaming interface.

File: logstash-core/src/main/java/org/logstash/common/Util.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Created by andrewvc on 12/23/16.
+ */
+public class Util {
+    // Modified from http://stackoverflow.com/a/11009612/11105
+    public static String sha256(String base) {
+        try{
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            byte[] hash = digest.digest(base.getBytes("UTF-8"));
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/common/Util.java]: Nice! Did not know...

File: logstash-core/src/main/java/org/logstash/common/Util.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            return bytesToHexString(hash);
+        } catch(Exception ex){
+            throw new RuntimeException(ex);
+        }
+    }
+
+    public static String bytesToHexString(byte[] bytes) {
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/common/Util.java]: Hmmm, that's awesome, but I don't know if I want to tie this to the java XML processing functions.

File: logstash-core/src/main/java/org/logstash/config/ir/Pipeline.java [hunk 12]
```diff
@@ -30,32 +30,23 @@ public SpecialVertex getFilterOut() {
...
         // Now we connect the leaves (and partial leaves) of the graph
         // which should all be filters (unless no filters are defined)
         // to the special filterOut node
         filterOut = new SpecialVertex(SpecialVertex.Type.FILTER_OUT);
-        graph.threadLeavesInto(filterOut);
+        tempGraph = tempGraph.chain(filterOut);
 
         // Finally, connect the filter out node to all the outputs
-        graph.threadLeavesInto(outputSection);
+        this.graph = tempGraph.chain(outputSection);
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/Pipeline.java]: Yeah, took a while to get the naming right.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java [hunk 13]
```diff
@@ -1,182 +1,294 @@
...
+        // If this belongs to another graph use a copy
+        if (v.getGraph() != null && v.getGraph() != this) {
+            throw new InvalidIRException("Attempted to add vertex already belonging to a graph!");
+        }
+
+        v.setGraph(this);
+
         this.vertices.add(v);
         this.refresh();
+    }
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java]: YES! Good catch!

Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java]: Now fixed

File: logstash-core/src/main/java/org/logstash/config/ir/graph/Vertex.java [hunk 14]
```diff
@@ -94,14 +93,87 @@ public boolean hasOutgoingEdges() {
...
+        return this.getGraph().getIncomingEdges(this).stream();
     }
 
     public Stream<Edge> outgoingEdges() {
-        return outgoingEdges.stream();
+        return this.getGraph().getOutgoingEdges(this).stream();
+    }
+
+    public Stream<Vertex> ancestors() {
+        // Without all the distinct calls this can be slow
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/Vertex.java]: Oh, I previously had a different algorithm. I'll kill this comment.

File: logstash-core/src/main/java/org/logstash/config/ir/graph/algorithms/ShortestPath.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+import org.logstash.config.ir.graph.Vertex;
+
+import java.util.*;
+import java.util.stream.Stream;
+
+/**
+ * Created by andrewvc on 1/5/17.
+ * This class isn't currently used anywhere, I wrote it for some code that is now removed, however, I'm sure it will be
+ * useful shortly, so we should hold onto it for a while.
+ */
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/config/ir/graph/algorithms/ShortestPath.java]: Well, I have tests for it, so I'm not too worried about that TBH.

**Comment by ph** [comment]
Lets close this PR since its merge in the feature branch.
